allennlp/models/constituency_parser.py

```diff
@@ -0,0 +1,355 @@
+        """
+        Constructs a tree given the scored spans.
+        """
+        all_predictions = output_dict['class_probabilities'].cpu().data
+        all_spans = output_dict["spans"].cpu().data
+        no_label_id = self.vocab.get_token_index("NO-LABEL", "labels")
+
+        all_sentences = output_dict["tokens"].data
+        sentence_lengths = get_lengths_from_binary_sequence_mask(output_dict["token_mask"]).data
+
+        trees = []
+        for batch_index, (predictions, spans, sentence_ids) in enumerate(zip(all_predictions,
+                                                                             all_spans,
+                                                                             all_sentences)):
+            sentence: List[str] = [self.vocab.get_token_from_index(index, "tokens") for
+                                   index in sentence_ids[:sentence_lengths[batch_index]]]
+
+            selected_spans = []
+            for prediction, span in zip(predictions, spans):
+                start, end = span
+                no_label_prob = prediction[no_label_id]
+                label_prob, label_index = torch.max(prediction, -1)
+
+                # Does the span have a label != NO-LABEL or is it the root node?
+                # If so, include it in the spans that we consider.
+                if int(label_index) != no_label_id or (start == 0 and end + 1 == len(sentence)):
+                    selected_spans.append({
+                            "start": int(start),
+                            # Switch to exclusive span ends to make
+                            # recursive tree constuction easier.
+                            "end": float(end) + 1,
+                            "label_prob": float(label_prob),
+                            "no_label_prob": float(no_label_prob),
+                            "label_index": float(label_index)
+                    })
+
+            # The spans we've selected might overlap, which causes problems when we try
+            # to construct the tree as they won't nest properly.
+            consistent_spans = self.resolve_overlap_conflicts_greedily(selected_spans)
+
+            spans_to_labels = {(int(span["start"]), int(span["end"])):
+                               self.vocab.get_token_from_index(int(span["label_index"]), "labels")
+                               for span in consistent_spans}
+            trees.append(self.construct_tree_from_spans(spans_to_labels, sentence))
+
+        output_dict["trees"] = trees
+        return output_dict
+
+    @staticmethod
+    def resolve_overlap_conflicts_greedily(chosen_spans: List[Dict[str, float]]) -> List[Dict[str, float]]:
+        """
+        Given a set of spans, removes spans which overlap by evaluating the difference
+        in probability between one being labeled and the other explicitly having no label
+        and vice-versa. The worst case time complexity of this method is ``O(k * n^4)`` where ``n``
```

Comment:
Saying this is `n^4` made me think there'd be four nested loops, but a factor of `n^2` here just comes from the fact that we're dealing with spans.  You might want to make `n` be the number of spans, not the length of the sentence.